File: csharp/Demo/Patterns.cs
using System;
using FluentAssertions;
using Xunit;

namespace Demo;

public class Patterns
{
    private interface IResult
    {
    }

    private record Ok<T>(T Data) : IResult;

    private record Failed : IResult;

    [Fact]
    public void Match()
    {
        IResult o = new Ok<int>(3);
        if (o is Ok<int> d)
        {
            d.Data.Should().Be(3);
        }
    }

    private static int Do(IResult result) => 
        result switch
        {
            Ok<int> o when o.Data % 2 == 0 => 1,
            Ok<bool> { Data: true } => 2,
            Failed => 3,
            _ => 4
        };

    [Fact]
    public void Switch()
    {
        var result = new Ok<int>(3);
        Do(result)
            .Should()
            .Be(4);
    }
}

File: kotlin/src/test/kotlin/Patterns.kt
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

class Patterns {
    interface Result
    data class Ok<T>(val data: T) : Result
    data class Failed(val message: String) : Result

    private fun match(result: Result) = when (result) {
        is Ok<*> -> { // type erasure
            when (result.data) { // Any?
                is Int ->  if (result.data % 2 == 0) 1 else 4 // smart cast
                is Boolean -> when {
                    result.data -> 2
                    else -> 4
                }
                else -> 4
            }
        }
        is Failed -> 3
        else -> 4
    }

    @Test
    fun odd() =
        assertEquals(4, match(Ok(3)))

    @Test
    fun even() =
        assertEquals(1, match(Ok(2)))

    @Test
    fun isTrue() =
        assertEquals(2, match(Ok(true)))

    @Test
    fun isFalse() =
        assertEquals(4, match(Ok(false)))

    @Test
    fun failed() =
        assertEquals(3, match(Failed("because")))

    @Test
    fun other() =
        assertEquals(4, match(object: Result {}))
}